enhancements/automated-service-ca-rotation.md

```diff
@@ -0,0 +1,246 @@
+  long as the pre-rotation signing CA remains valid. The following should be
+  true:
+  - A serving cert, from the new CA, validates against the pre-rotation CA
+    bundle.
+  - A serving cert, from the old CA, validates against the post-rotation CA
+    bundle.
+
+- After rotation:
+  - Serving cert secrets should be updated with serving certs generated from the new CA.
+  - Configmaps annotated for CA bundle injection should be updated with the new CA bundle.
+  - API services annotated for CA bundle injection should be updated with the new CA bundle.
+
+- Support forced rotation
+  - Rotation triggered by an API change rather than the validity bounds of the CA
+
+- Support forced rotation with a custom validity bounds to support emergency rotation
+  - i.e. support minimizing the grace period for trusting the pre-rotation CA
+
+### Non-Goals
+
+- Support forced rotation using an external CA and signing key. This could be
+  useful for admins to plug in their own intermediate CA and have all serving
+  certs issued by it, and also for future integration efforts.
+
+- Implement support for [hitless
+rotation](https://diogomonica.com/2017/01/11/hitless-tls-certificate-rotation-in-go/)
+so that consumers of serving certs and CA bundles can trivially support automated
+refresh. This is already supported by
+[controller-runtime](https://github.com/kubernetes-sigs/controller-runtime/pull/421)
+and the functionality could be separately extracted for operator reuse.
+
+## Proposal
+
+- Detect when the current CA certificate has passed 1/2 of its validity bounds
+  (i.e. 6 months has passed of a 1 year validity) and generate a new CA.
+  - The new CA should have the same `Subject.CommonName` as the current CA to
+    ensure that trust chaining is possible.
+  - This provides a 6 month grace period for consumers of serving certs and ca
+    bundles to be refreshed.
+
+- Generate an intermediate CA certificate with the same public key as the new CA
+  but signed with the private key of the current CA.
+  - This intermediate certificate should be included with newly-generated serving
+    certificates, and serves to bridge trust between the CA bundle of an
+    unrefreshed client (validating trust with the old CA bundle) and a service
+    serving with a cert generated from the new CA.
+
+- Generate an intermediate CA certificate with the same public key as the current CA
+  but signed with the private key of the new CA.
+  - This intermediate certificate should be included in the the new CA
+    bundle along with the new CA's primary certificate, and serves to
+    bridge trust between an unrefreshed server (serving with a cert
+    generated from the previous CA) and a refreshed client (validating
+    trust with the new CA bundle).
+
+- Compare the `AuthorityKeyId` of a serving cert with the `SubjectKeyId` of the
+  current CA to determine whether the serving cert was issued by the current CA.
```

Comment:
This seems like this check would not work if the old generation of certs are in their current state (with no auth ID), only for generations beyond one rotation. Instead you might doi a x509.Verify() with the serving cert (without its intermediate) against the newest CA. If the verify fails, then renew. 

There is still a good reason to have the key IDs included, and that is to help clients build the trust chain in the cross-signed scenario.

Reply:
I'm not sure why verify would be required as per the item on line 125. Also, at least in golang, key identity chaining is [subordinate](https://github.com/golang/go/blob/master/src/crypto/x509/cert_pool.go#L76) to common name chaining, so key ids don't appear to play a role in build ing a trust chain in a world where the same common name is maintained across CAs.

Reply:
The verify I'm suggesting as a way to check that a current serving cert is issued by the current CA, which will work even when your current serving cert has no AuthorityKeyID (i.e., before the first rotation after upgrade).

With regards to client chaining, the code you linked says the opposite, where the AuthorityKeyID is checked first and then if there's none found then look at the Issuer, which is the behavior that I hoped golang would follow :) The point of the Sub/AuthKeyIDs is to give clients a more reliable mechanism for following an issuance chain without having to rely on mandating unique Subject/Issuer fields (which are purely informational, and can even be blank). OpenSSL clients also prefer AuthorityKeyIDs when present over Subject/Issuer comparison.

Reply:
Whoops, you are correct re: key identity chaining having presence over name-based chaining.  I'm not sure what I was thinking before.

I think you're right that for a general-purpose solution formally verifying issuance might make sense. For detecting whether a serving cert was issued by the service CA, though, I'm not sure it's worth it. There are no intermediate certs involved, so the serving cert was either issued by the current CA or it wasn't, and the operator is the sole arbiter of how certificates are created.

As proposed, if the current CA has `SubjectKeyId` set, then issuance will continue to be determined by common name. If the current CA has `SubjectKeyId` set, then key identity chaining will be used, and serving certs lacking `AuthorityKeyId` will be regenerated. I'm not sure it's worth putting much effort into switching to certificate verification from name-based chaining because the the lifespan of a legacy cert (not having key identity set) is likely to be limited once a cluster is upgraded to a version (>= 4.3) supporting automated rotation.

Reply:
After writing that, I think the primary reason to take the short-cut of determining issuance based on name or key identity is to avoid the potential cryptographic overhead of `Verify`.

Reply:
Agreed, the potential for overhead is there. Using the key IDs like you suggest is probably good enough given the upgrade path.